### PR TITLE
feat: Provably build derived op blocks

### PR DESCRIPTION
**Pull request by hashcashier**
This PR amends the optimism block derivation process to verifiably build (using composition) the block associated with each batch before choosing whether to drop or adopt the batch as the new op head. The transaction data extracted during derivation is used directly as input for the block building process, which is executed as its own zkvm session whose receipt is pulled as an assumption for derivation in via composition.

### CODE REVIEW

**Comment by intoverflow** [comment]
Do we want to execute Op transactions in the derive process? I was thinking that we'd use Zeth for that, as described [here](https://github.com/ethereum-optimism/ecosystem-contributions/issues/61#issuecomment-1858698611).

**Comment by intoverflow** [comment]
Here's my only hesitation:

The derive process (without tx execution) is very expensive. In the linked Op Rfp comment, running derive for 10 blocks is >1B cycles. It's expensive for two reasons:

1. It must read a bunch of Eth blocks, and
2. Batch data tends to be "clumped up:" you can have several Eth blocks in a row with no useful data, then suddenly a bunch of batches all appear at the same time.

Because of this, it makes sense to derive several blocks in the same proof (as this re-uses the work needed to read so many Eth blocks).

But if the derive process also executes transactions, we run into problems with having several blocks derived in 1 proof. (At least, that's what happened when we tried this a few months ago; I will try again with this PR to see if the situation has changed.)

Tldr: including the tx execution in the derive process might increase our total cycle count, since it might restrict us to only deriving 1 block per proof, which prevents us from being able to amorize the costly derive process across several blocks.

File: primitives/src/trie/mpt.rs [hunk 1]
```diff
@@ -71,6 +71,15 @@ pub enum Error {
     LegacyRlp(#[from] DecoderError),
 }
 
+impl From<B256> for MptNode {
+    fn from(value: B256) -> Self {
+        MptNode {
+            data: MptNodeData::Digest(value),
```
Comment (Wollac) [inline primitives/src/trie/mpt.rs:77]: For the EMPTY_ROOT this should be Null instead.

There is already `node_from_digest` which does that, so we should consider consolidating.

However, it is important to note that sometimes the API returns B256::ZERO when it should instead be EMPTY_ROOT. So to be on the safe side we should:
```rust
    match digest {
        EMPTY_ROOT | B256::ZERO => MptNode::default(),
        _ => MptNodeData::Digest(digest).into(),
    }
```

File: guests/eth-block/src/main.rs [hunk 2]
```diff
@@ -26,10 +26,15 @@ pub fn main() {
     // Read the input previous block and transaction data
     let input = env::read();
     // Build the resulting block
-    let (header, state) = EthereumStrategy::build_from(&ETH_MAINNET_CHAIN_SPEC, input)
+    let mut output = EthereumStrategy::build_from(&ETH_MAINNET_CHAIN_SPEC, input)
         .expect("Failed to build the resulting block");
-    // Output the resulting block's hash to the journal
-    env::commit(&header.hash());
+    // Abridge successful construction results
+    if let Some(trie_root) = output.compress_state() {
```
Comment (Wollac) [inline guests/eth-block/src/main.rs:32]: `compress_state` replaces the state in the output with just its root and returns the original state.

In our current state/preflight code we use the following naming scheme:
- ..._root denotes the root hash
-- ..._root_node denotes the root MptNode with all its children

Thus, this should here be called either `state_root_node` or just `state` as it is not just the root hash

File: lib/src/optimism/mod.rs [hunk 3]
```diff
@@ -236,119 +282,217 @@ impl<D: BatcherDb> DeriveMachine<D> {
...
+                    .get_full_eth_block(op_batch.0.epoch_num)
+                    .context("eth block not found")?
+                    .block_header
+                    .mix_hash;
+
+                // From the spec:
+                // The first transaction MUST be a L1 attributes deposited transaction,
+                // followed by an array of zero-or-more user-deposited transactions.
+                let l1_attributes_tx = self.derive_l1_attributes_deposited_tx(&op_batch);
+                // TODO: revise that skipping undecodable transactions is part of spec
```
Comment (Wollac) [inline lib/src/optimism/mod.rs:304]: Important point, we might want to check what OP-geth is doing with invalid transactions

File: lib/src/optimism/mod.rs [hunk 4]
```diff
@@ -236,119 +282,217 @@ impl<D: BatcherDb> DeriveMachine<D> {
...
+                            state_root: Default::default(),
+                            transactions_root: Default::default(),
+                            receipts_root: Default::default(),
+                            logs_bloom: Default::default(),
+                            difficulty: Default::default(),
+                            number: 0,
+                            gas_used: Default::default(),
+                            nonce: Default::default(),
+                            base_fee_per_gas: Default::default(),
+                            withdrawals_root: None,
```
Comment (Wollac) [inline lib/src/optimism/mod.rs:393]:
```suggestion
                            ..Default::default()
```

File: lib/src/builder/mod.rs [hunk 5]
```diff
@@ -28,22 +31,43 @@ use crate::{
...
 pub struct BlockBuilder<'a, D, E: TxEssence> {
     pub(crate) chain_spec: &'a ChainSpec,
-    pub(crate) input: Input<E>,
+    pub(crate) input: BlockBuildInput<E>,
     pub(crate) db: Option<D>,
     pub(crate) header: Option<Header>,
+    pub db_drop_destination: Option<DatabaseRescue<D>>,
+}
+
+// This implementation allows us to recover data during erroneous block builds on the host
```
[the remark this answers is not in the document]
Comment (hashcashier) [reply lib/src/builder/mod.rs:59]: later would be better as this I recall this looked like it needed a lot of refactoring instead

File: lib/src/builder/mod.rs [hunk 6]
```diff
@@ -111,13 +145,41 @@ pub trait BlockBuilderStrategy {
...
-    ) -> Result<(Header, MptNode)> {
-        BlockBuilder::<MemDb, Self::TxEssence>::new(chain_spec, input)
-            .initialize_database::<Self::DbInitStrategy>()?
-            .prepare_header::<Self::HeaderPrepStrategy>()?
-            .execute_transactions::<Self::TxExecStrategy>()?
-            .finalize::<Self::BlockFinalizeStrategy>()
+        input: BlockBuildInput<Self::TxEssence>,
+    ) -> Result<BlockBuildOutput> {
+        // Database initialization failure does not mean the block is faulty
+        let input_hash = input.partial_hash();
```
[the remark this answers is not in the document]
Comment (hashcashier) [reply lib/src/builder/mod.rs:151]: yeah just adding a todo because it's a bit of a refactor to not have the builder own the input or return that only on error. luckily it's a sha256

File: lib/src/optimism/mod.rs [hunk 7]
```diff
@@ -236,122 +281,214 @@ impl<D: BatcherDb> DeriveMachine<D> {
...
+                    .0
+                    .transactions
+                    .iter()
+                    .map(|tx| Transaction::decode_strict(&mut tx.as_ref()))
+                    .filter_map(|tx| tx.ok())
+                    // We always assume that chain id exists here
+                    .map(
+                        |mut tx: Transaction<OptimismTxEssence>| match &mut tx.essence {
+                            OptimismTxEssence::Ethereum(EthereumTxEssence::Legacy(essence)) => {
+                                if essence.chain_id.is_none() {
```
Comment (Wollac) [inline lib/src/optimism/mod.rs:314]: This must not be needed, the chain ID should be set during RLP decoding

File: primitives/src/transactions/ethereum.rs [hunk 8]
```diff
@@ -53,6 +58,43 @@ pub struct TxEssenceLegacy {
...
+    pub r: U256,
+    pub s: U256,
+}
+
+impl SignedDecodable<TxSignature> for TxEssenceLegacy {
+    fn decode_signed(buf: &mut &[u8]) -> alloy_rlp::Result<(Self, TxSignature)> {
+        let signed_essence = TxEssenceLegacyTxSignature::decode(buf)?;
+        Ok((
+            Self {
+                chain_id: None,
```
Comment (Wollac) [inline primitives/src/transactions/ethereum.rs:81]: This needs to be set from the signature for EIP-155 transactions.

**Review by Wollac — CHANGES_REQUESTED** [review]

File: lib/src/optimism/mod.rs (hunk 7 above)
Comment (Wollac) [reply lib/src/optimism/mod.rs]: done

File: primitives/src/transactions/ethereum.rs (hunk 8 above)
Comment (Wollac) [reply primitives/src/transactions/ethereum.rs]: done

**Review by Wollac — APPROVED** [review]
The transaction RLP-Encoding/Decoding is getting more and more ugly. But we can address that in a different PR or hope that alloy will implement it for us.

File: lib/src/optimism/mod.rs (hunk 3 above)
Comment (Wollac) [reply lib/src/optimism/mod.rs:306]: OK, I did some research on this:
- This is analog to Ethereum, the CL creates the `PayloadAttribute` and passes it to the EL to create the `ExecutionPayload`, essentially what we call block input (maybe we should consider renaming)
- The `ExecutionPayload` contains the RLP-encoded transactions, if they cannot be decoded the block is invalid.
- For OP the `PayloadAttribute` also contains transactions, that must then be included by the EL.
- If the `PayloadAttribute` contains invalid transactions, the entire request is invalid and we essentially have a consensus failure.
- I am not sure what mechanisms are there to recover from such a failure.

This means that undecodable transactions are **not** skipped, but they must all be valid.

However, I guess this means that we do not really need the RLP-decoding and could instead use the approach from before that downloads the transactions from RPC and just checks that the trie hash is correct. This could simplify our code dramatically.

File: lib/src/optimism/mod.rs [hunk 9]
```diff
@@ -236,122 +284,210 @@ impl<D: BatcherDb> DeriveMachine<D> {
...
-                            .context("eth block not found")?
-                            .block_header
-                            .clone();
+                match op_block_output {
+                    BlockBuildOutput::SUCCESS {
+                        hash: new_block_hash,
+                        head: new_block_head,
+                        ..
+                    } => {
+                        // Verify that the built op block matches the payload attributes of the
```
Comment (Wollac) [inline lib/src/optimism/mod.rs:418]: Why do we need to do that?
If we cannot trust the Input of the proof, then I guess we'd need to verify everything including the state input etc.

File: lib/src/builder/mod.rs [hunk 10]
```diff
@@ -111,13 +146,42 @@ pub trait BlockBuilderStrategy {
...
+        let prepared = match initialized.prepare_header::<Self::HeaderPrepStrategy>() {
+            Ok(builder) => builder,
+            Err(_) => {
+                return Ok(BlockBuildOutput::FAILURE {
+                    state_input_hash: input_hash.into(),
+                })
+            }
+        };
+
+        // Transaction execution errors mean a faulty block
```
Comment (Wollac) [inline lib/src/builder/mod.rs:167]: For actual error distinction, we would need to properly define a `ThisError' or something. The execution layer may return errors that are not failures, but other rust errors.

File: lib/src/optimism/mod.rs (hunk 3 above)
Comment (Wollac) [reply lib/src/optimism/mod.rs:306]: With this in mind, I am wondering wether we actually need to prove the block failure... This would also be some kind of consensus error and I don't think that skipping to the next batch is the designed behavior for such a case. Maybe we should just leave it at stopping at consensus errors.

**Review by Wollac — COMMENTED** [review]
Maybe we should discuss those considerations.

File: lib/src/optimism/mod.rs (hunk 3 above)
Comment (Wollac) [reply lib/src/optimism/mod.rs:306]: Funnily enough, op-geth neither returns a state reset nor an invalid attribute error, so i guess it would just stop in this case.

Comment (hashcashier) [reply lib/src/optimism/mod.rs:306]: there could be other problems with the block that could lead to dropping its batch, which needs to be justified with a proof.

File: lib/src/optimism/mod.rs (hunk 9 above)
Comment (hashcashier) [reply lib/src/optimism/mod.rs:418]: it'd have to be part of the success output as well which I think isn't a bad idea

File: lib/src/builder/mod.rs (hunk 10 above)
Comment (hashcashier) [reply lib/src/builder/mod.rs:167]: I'm not sure we can properly maintain a definitive distinction at that level
